### PR TITLE
Existing GameBinder.cs hooks implemented

### PR DESCRIPTION
**Pull request by RenderBr**

### CODE REVIEW

File: src/Starlight.Server/Hooks/HookCaller.cs [hunk 1]
```diff
@@ -65,6 +68,83 @@ public async Task<HandleResult> OnQuickStachAsync(OnQuickStackArgs args)
...
+
+
+        public async Task<HandleResult> OnStatueSpawnAsync(OnStatueSpawnArgs args)
+        {
+            foreach (var resolver in _resolvers)
+            {
+                var handle = await resolver.OnStatueSpawnAsync();
+
+                if (handle.Handled)
+                    break;
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookCaller.cs:132]:
```suggestion
                    return handle;
```

File: src/Starlight.Server/Hooks/HookCaller.cs [hunk 2]
```diff
@@ -65,6 +68,83 @@ public async Task<HandleResult> OnQuickStachAsync(OnQuickStackArgs args)
             return HandleResult.Continue();
         }
 
+
+        private bool currentGameMenuState;
+        public async Task InvokeGameUpdateAsync()
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookCaller.cs:73]:
```suggestion
        public async Task OnGameUpdateAsync()
```

Inconsistent naming. Why are we adding `Invoke` here?

File: src/Starlight.Server/Hooks/HookCaller.cs [hunk 3]
```diff
@@ -65,6 +68,83 @@ public async Task<HandleResult> OnQuickStachAsync(OnQuickStackArgs args)
...
+        {
+            if(currentGameMenuState != Terraria.Main.gameMenu)
+            {
+                currentGameMenuState = Terraria.Main.gameMenu;
+
+                if (Terraria.Main.gameMenu)
+                    await InvokeGameWorldDisconnect();
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookCaller.cs:80]:
```suggestion
                    await OnGameWorldDisconnect();
```

File: src/Starlight.Server/Hooks/HookCaller.cs [hunk 4]
```diff
@@ -65,6 +68,83 @@ public async Task<HandleResult> OnQuickStachAsync(OnQuickStackArgs args)
...
+                else
+                    await InvokeGameWorldConnect();
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookCaller.cs:82]:
```suggestion
                    await OnGameWorldConnect();
```

File: src/Starlight.Server/Hooks/HookCaller.cs [hunk 5]
```diff
@@ -65,6 +68,83 @@ public async Task<HandleResult> OnQuickStachAsync(OnQuickStackArgs args)
...
+                            
+            }
+
+            foreach (var resolver in _resolvers)
+            {
+                var handle = await resolver.InvokeGameUpdateAsync();
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookCaller.cs:88]:
```suggestion
                var handle = await resolver.OnGameUpdateAsync();
```

File: src/Starlight.Server/Hooks/HookCaller.cs [hunk 6]
```diff
@@ -65,6 +68,83 @@ public async Task<HandleResult> OnQuickStachAsync(OnQuickStackArgs args)
...
+            return;
+
+        }
+
+
+        public async Task InvokePostGameUpdateAsync()
+        {
+            foreach (var resolver in _resolvers)
+            {
+                var handle = await resolver.InvokePostGameUpdateAsync();
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookCaller.cs:103]:
```suggestion
                var handle = await resolver.OnPostGameUpdateAsync();
```

File: src/Starlight.Server/Hooks/HookCaller.cs [hunk 7]
```diff
@@ -65,6 +68,83 @@ public async Task<HandleResult> OnQuickStachAsync(OnQuickStackArgs args)
...
+                if (handle.Handled)
+                    break;
+            }
+
+            return;
+
+        }
+
+
+        public async Task InvokePostGameUpdateAsync()
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookCaller.cs:99]:
```suggestion
        public async Task OnPostGameUpdateAsync()
```

File: src/Starlight.Server/Hooks/HookCaller.cs [hunk 8]
```diff
@@ -1,4 +1,7 @@
 ﻿using CSF;
+using IL.Terraria;
+using MongoDB.Driver;
+using System.Reflection.Metadata.Ecma335;
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookCaller.cs:4]:
```suggestion
```

accident?

File: src/Starlight.Server/Hooks/HookCaller.cs [hunk 9]
```diff
@@ -1,4 +1,7 @@
 ﻿using CSF;
+using IL.Terraria;
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookCaller.cs:2]:
```suggestion
using Terraria;
```

File: src/Starlight.Server/Hooks/HookCaller.cs [hunk 10]
```diff
@@ -1,4 +1,7 @@
...
+using MongoDB.Driver;
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookCaller.cs:3]:
```suggestion
```

File: src/Starlight.Server/Hooks/Binders/Implementations/GameBinder.cs [hunk 11]
```diff
@@ -29,18 +29,19 @@ public static void Bind(HookCaller caller)
...
         private static void OnUpdate(On.Terraria.Main.orig_Update orig, Terraria.Main instance, GameTime gameTime)
         {
-            _ = _caller.InvokeGameUpdateAsync();
+
+             _ = _caller.InvokeGameUpdateAsync();
 
             orig(instance, gameTime);
 
-            _ = _caller.InvokeGamePostUpdateAsync();
+            _ = _caller.InvokePostGameUpdateAsync();
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/Binders/Implementations/GameBinder.cs:37]:
```suggestion
            _ = _caller.OnPostGameUpdateAsync();
```

File: src/Starlight.Server/Hooks/Binders/Implementations/GameBinder.cs [hunk 12]
```diff
@@ -29,18 +29,19 @@ public static void Bind(HookCaller caller)
 
         private static void OnUpdate(On.Terraria.Main.orig_Update orig, Terraria.Main instance, GameTime gameTime)
         {
-            _ = _caller.InvokeGameUpdateAsync();
+
+             _ = _caller.InvokeGameUpdateAsync();
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/Binders/Implementations/GameBinder.cs:33]:
```suggestion
             _ = _caller.OnGameUpdateAsync();
```

File: src/Starlight.Server/Hooks/HookResolver.cs [hunk 13]
```diff
@@ -86,10 +86,40 @@ public virtual Task<HandleResult> OnChatBroadcastAsync(OnBroadcastArgs args)
...
+        {
+            return Continue();
+        }
+
+        public virtual Task<HandleResult> InvokeGameUpdateAsync()
+        {
+            return Continue();
+        }
+
+        public virtual Task<HandleResult> InvokePostGameUpdateAsync()
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookResolver.cs:104]:
```suggestion
        public virtual Task<HandleResult> OnPostGameUpdateAsync()
```

File: src/Starlight.Server/Hooks/HookResolver.cs [hunk 14]
```diff
@@ -86,10 +86,40 @@ public virtual Task<HandleResult> OnChatBroadcastAsync(OnBroadcastArgs args)
...
+        {
+            return Continue();
+        }
+
+        public virtual Task<HandleResult> OnHardmodeTilePlaceAsync(OnHardmodeTilePlaceArgs args)
+        {
+            return Continue();
+        }
+
+        public virtual Task<HandleResult> InvokeGameUpdateAsync()
```
Comment (Rozen4334) [inline src/Starlight.Server/Hooks/HookResolver.cs:99]:
```suggestion
        public virtual Task<HandleResult> OnGameUpdateAsync()
```

**Review by Rozen4334 — APPROVED** [review]
lgtm

**Review by Rozen4334 — CHANGES_REQUESTED** [review]
Why is it invoke again grrr

**Comment by RenderBr** [comment]
> Why is it invoke again grrr

fixed. sorry babe dont growl at me god

**Comment by RenderBr** [comment]
> Why is it invoke again grrr

why do you do this every time 😢

**Comment by Rozen4334** [comment]
Force merging as naming is inappropriate and I want to continue on hooks branch

**Pull request merged**